### PR TITLE
tianyu dev: spade masker

### PR DESCRIPTION
**Pull request by vict0rsch**

### CODE REVIEW

**Comment by vict0rsch** [comment]
@tianyu-z Can you turn that into a `MaskSpadeDecoder` class? instead of `G.spade`?

**Comment by vict0rsch** [comment]
check that if use spade for  m: s and d are in the tasks

**Comment by vict0rsch** [comment]
```
class MaskSpadeDecoder(nn.Module):

	def __init__(self, ...):
		
		self.mask_conv = nn.Conv2d(self.final_nc, 1, 3, padding=1)


    def forward(self, z, cond):
		y = self.spade_1(z, cond)
		y = self.up(y)
		y = self.spade_2(z, cond)
		y = self.up(y)
		y = self.spade_3(z, cond)
		y = self.up(y)
		return self.mask_conv(y)
```

**Comment by vict0rsch** [comment]
@tianyu-z I have a slight reservation in your implementation of the forward pass in case of dlv3 resnets.

My intuition would have been to somehow concatenate and project the low and high level feature to keep a single spade line of processing. What you have implemented makes the spade block treat in 2 different parallel paths each features but I'm not sure it makes sense that it should know what kind of features it gets and how it should treat it differently. Do you see what I mean?

I'd have done something simpler like in case of resnet as I have implemented in the `BaseDecoder` with the `low_level_conv` and `merge_feats_conv`. 

If you don't agree with this path feel free to discuss as always :) !

File: omnigan/trainer.py [hunk 1]
```diff
@@ -1569,17 +1613,21 @@ def masker_s_loss(self, x, z, target, domain, for_="G"):
                         full_loss += gp_loss
                     else:
                         raise NotImplementedError
-
+        if self.opts.gen.m.use_spade:
+            return full_loss, softmax(pred, dim=1).detach()
```
Comment (vict0rsch) [inline omnigan/trainer.py:1617]: @tianyu-z I think it would be clearer if we kept a constant API: `masker_X_loss` should always return a tuple `loss, pred` it's just that if we don't use spade for `m` then we'll just ignore the second item of the tuple

File: omnigan/trainer.py [hunk 2]
```diff
@@ -1569,17 +1613,21 @@ def masker_s_loss(self, x, z, target, domain, for_="G"):
...
         assert domain in {"r", "s"}
         self.assert_z_matches_x(x, z)
         assert x.shape[0] == target.shape[0] if target is not None else True
         full_loss = torch.tensor(0.0, device=self.device)
         # ? output features classifier
-        pred_logits = self.G.mask(z=z, sigmoid=False)
+        if self.opts.gen.m.use_spade:
+            pred_logits = self.G.decoders["m"](z, cond)
+        else:
+            pred_logits = self.G.decoders["m"](z)
```
Comment (vict0rsch) [inline omnigan/trainer.py:1630]: I updated the API to the non-spade masker so that we don't have to care about this

**Comment by vict0rsch** [comment]
@tianyu-z I did a quick pass to kinda fix and make uniform the overall API but not the SPADE decoder itself as i want to have your opinion
